### PR TITLE
Multi challenge rengo with types

### PR DESCRIPTION
**Pull request by GreenAsJade**
Fixes the problem of not being able to manage multiple outstanding rengo correspondence challenges

## Proposed Changes

 - Rengo live challenges are one only at a time, and the management is done in the automatch container
 - Rengo correspondence challenges can be managed independently from automatch container: a pane opens up in-place in the Rengo challenge list
 -  Live & Corre rengo challenges are more visually separated, due to their different controls.
 - Challenges are typed in Play.tsx and elsewhere, courtesy of @benjaminpjones PR that does this.
 - @shinuito's Safari 13.s fix is incorporated

### CODE REVIEW

**Comment by anoek** [comment]
Excellent, will be good to have a better solution for correspondence games.

Would it be possible to open that not in the same column as the buttons? Doing so makes it look like this:
image

File: src/views/Play/Play.tsx [hunk 1]
```diff
@@ -348,32 +355,31 @@ export class Play extends React.Component<{}, PlayState> {
...
-    ownRengoChallengePending = (): any => {
-        const orcp = this.state.rengo_list.find((c) => (c.user_challenge));
+    ownRengoChallengesPending = (): Challenge[] => { // multiple correspondence are possible, plus one live
+        const orcp = this.state.rengo_list.filter((c) => (c.user_challenge));
+        //console.log("own rcp", orcp);
         return orcp;
     };
 
-    joinedRengoChallengePending = (): any => {
+    joinedRengoChallengesPending = (): any[] => { // multiple correspondence are possible, plus one live
```
Comment (benjaminpjones) [inline src/views/Play/Play.tsx:375]: Oops I missed one - looks like `any[]` can be `Challenge[]`

Comment (GreenAsJade) [reply src/views/Play/Play.tsx:375]: Heh heh, I'm the one that missed it.  I had to "chose ours" during the merge for this part of the code, I missed going back to that one.

**Comment by GreenAsJade** [comment]
@anoek I'll see if I can merge the cells or something like that, so it doesn't insert itself that way.

File: src/components/RengoManagementPane/RengoManagementPane.tsx [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+interface Events {
+}
+
+interface RengoManagementPaneProperties {
+    user_id: number;
+    challenge_id: number;
+    rengo_challenge_list: any[];
+
+    startRengoChallenge: (challenge: any) => void;
```
Comment (benjaminpjones) [inline src/components/RengoManagementPane/RengoManagementPane.tsx:30]: I guess these can be typed now too :)

File: src/components/RengoManagementPane/RengoManagementPane.tsx [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import * as React from "react";
+
+import {_, pgettext, interpolate} from "translate";
+
+interface Events {
```
Comment (benjaminpjones) [inline src/components/RengoManagementPane/RengoManagementPane.tsx:22]: Unused interface

File: src/components/RengoManagementPane/RengoManagementPane.tsx [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import * as React from "react";
+
+import {_, pgettext, interpolate} from "translate";
```
Comment (benjaminpjones) [inline src/components/RengoManagementPane/RengoManagementPane.tsx:20]: `interpolate` is unused

File: src/components/RengoTeamManagementPane/RengoTeamManagementPane.tsx [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+import * as React from "react";
+import { _, pgettext, interpolate } from "translate";
+
+import * as data from "data";
+
+import { Player } from "Player";
+
+interface RengoTeamManagementPaneProps {
+    challenge_list: any[];
```
Comment (benjaminpjones) [inline src/components/RengoTeamManagementPane/RengoTeamManagementPane.tsx:26]: `Challenge[]`

File: src/components/RengoTeamManagementPane/RengoTeamManagementPane.tsx [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+    challenge_id: number;
+    assignToTeam: (
+        player_id: number,
+        team: string,
+        challenge: any,
```
Comment (benjaminpjones) [inline src/components/RengoTeamManagementPane/RengoTeamManagementPane.tsx:31]: Same here

File: src/components/RengoTeamManagementPane/RengoTeamManagementPane.tsx [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import * as React from "react";
+import { _, pgettext, interpolate } from "translate";
+
+import * as data from "data";
```
Comment (benjaminpjones) [inline src/components/RengoTeamManagementPane/RengoTeamManagementPane.tsx:21]: Unused imports: `pgettext`, `interpolate`, `data`

File: src/models/challenges.d.ts [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+    "rules": import('../lib/types').RuleSet;
+    "width": number;
+    "height": number;
+    "challenger_color": "black" | "white" | "automatic";
+    "disable_analysis": true;
+    "time_control": import('../components/TimeControl').TimeControlTypes.TimeControlSystem;
+    "time_control_parameters": import('../components/TimeControl').TimeControl;
+    "time_per_move": number;
+    "rengo": boolean;
+    "rengo_nominees": any[];
```
Comment (benjaminpjones) [inline src/models/challenges.d.ts:48]: @GreenAsJade do you have the context on these `any` types?  I probably could have looked at an actual rengo challenge when building this, but if you already know what these are, that's even better!

File: src/views/Play/Play.tsx [hunk 9]
```diff
@@ -749,13 +737,58 @@ export class Play extends React.Component<{}, PlayState> {
...
+            (this.state.show_other_boardsize_challenges &&
+                (
+                    C.width !== C.height ||
+                    (C.width !== 19 && C.width !== 13 && C.width !== 9)
+                )
+            )
+        )
+    );
+
+    suspectChallengeIcon = (C) => (
```
Comment (benjaminpjones) [inline src/views/Play/Play.tsx:756]: ❤️

**Review by benjaminpjones — COMMENTED** [review]
Very excited to be able to join correspondence rengo and still do other stuff!

Left some minor comments (mostly about types) no need to address in this PR just wanted to point them out!

**Comment by GreenAsJade** [comment]
@anoek  the above commit deals with the issue you raised.

Challenge List is quite a hair-raising table, bending HTML in wierd and wonderful ways.

I decided to break Rengo out into it's own table, so I didn't have to interact with the rest of the table.

I think I had to make it an actual table to make colspan work, though TBH there was a ton of trial and error there so I couldn't tell you precisely what the minimal set of changes would be to make it work ... but it seems to work now.

File: src/views/Play/Play.tsx [hunk 10]
```diff
@@ -900,182 +919,156 @@ export class Play extends React.Component<{}, PlayState> {
...
         const assignment = team === 'rengo_black_team' ? 'assign_black' :
             team === 'rengo_white_team' ? 'assign_white' :
             'unassign';
 
-        this.setState({admin_pending: true});
-
         put("challenges/%%/team", challenge.challenge_id, {
             [assignment]: [player_id, ]  // back end expects an array of changes, but we only ever send one at a time.
         })
         .then(() => {
```
[the remark this answers is not in the document]
Comment (GreenAsJade) [reply src/views/Play/Play.tsx:1090]: 🤦‍♀️   of course! :)

File: src/models/challenges.d.ts (hunk 8 above)
Comment (GreenAsJade) [reply src/models/challenges.d.ts:48]: Player IDs

File: src/views/Play/Play.tsx (hunk 9 above)
Comment (GreenAsJade) [reply src/views/Play/Play.tsx:760]: Were you giggling at the capital C?   I think I copied that from previous code and could't stop doing it without being inconsistent :D

**Comment by anoek** [comment]
Awesome, thanks @GreenAsJade !

File: src/views/Play/Play.tsx (hunk 9 above)
Comment (benjaminpjones) [reply src/views/Play/Play.tsx]: No not giggling at anything! I'm just glad the cheat-alert text is no longer duplicated!

**Pull request merged**